README.md
# Consul Client

This Go Client library allows a service to register and deregister to consul. This library also allows to discover a service from consul by supplying app name and tags

## Getting Started

You can include this library in your glide.yaml/project by using:

`import github.com/usrlocalts/consul_client`

If you want to avoid entering username and password, in your glide.yaml, specifiy as follows:

```
- package: github.com/usrlocalts/consul_client
  repo: git@github.com:usrlocalts/consul_client.git
  vcs: git

```

## Contributing to the library

- `mkdir -p github.com/usrlocalts`
- Clone the repository `git clone git@github.com:usrlocalts/consul_client.git`
- `make test` to run tests

## Setting up go

This service runs on go.

- Install go
    - On OSX run `brew install go`.
    - Follow instructions on https://golang.org/doc/install for other OSes.
- Setup go
      - Make sure that the executable `go` is in your shell's path.
      - Add the following in your .zshrc or .bashrc: (where `<workspace_dir>` is the directory in
        which you'll checkout your code)
      - `make setup`
- Run Test
    make test

```
GOPATH=<workspace_dir>
export GOPATH
PATH="${PATH}:${GOPATH}/bin"
export PATH
```

## Usage

After you include the library in your application,

##### Create a client by doing the following:

```
func ConsulClientConfig() *consulClientConfig.Config {
	consulClientConfig := consulClientConfig.NewConfig()
	clientConfig.NodeID = "example-service"
    clientConfig.NodeIP = "example-ip"
    clientConfig.AppName = "example-service"
	return consulClientConfig
}

consulClient, err := client.NewConsul(ConsulClientConfig())


Assuming all these confgurations are in your config file

```

##### The default configurations when you do a `consulClientConfig.NewConfig()` are:

```
	Port:                      3000,
	ConsulAddress:             "localhost:8500",
	NodeIP:                    "localhost",
	ConsulCheckURL:            "http://localhost:3000/",
	ConsulCheckInterval:       "3s",
	ConsulTags:                []string{"grpc"},
	ConsulServiceCheckTimeout: 100,

```

##### You can edit these config by editing the configs:
 
```
	Port                      int
	ConsulAddress             string
	NodeID                    string
	NodeIP                    string
	AppName                   string
	ConsulCheckURL            string
	ConsulCheckInterval       string
	ConsulTags                []string
	ConsulServiceCheckTimeout int
	
	by
	
	clientConfig.Port = 3100

```

##### Methods available

```
consulClient.RegisterToConsul()
consulClient.DeRegisterFromConsul(serviceID)
consulClient.Discover("example-service", []string{"grpc"})
```

CHANGELOG.md
# consul_client CHANGELOG

1.0.0 [Shriram] 

* Supports Register and Deregister from Consul
* Supports Service Discovery

config/config_test.go
package config_test

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/usrlocalts/consul_client/config"
)

func TestConfig(t *testing.T) {
	config.NewConfig()
	assert.Equal(t, 3000, config.Port())
	assert.Equal(t, "localhost:8500", config.ConsulAddress())
	assert.Equal(t, "localhost", config.NodeIP())
	assert.Equal(t, "http://localhost:3000/", config.ConsulCheckURL())
	assert.Equal(t, "3s", config.ConsulCheckInterval())
	assert.Equal(t, []string{"grpc"}, config.ConsulTags())
	assert.Equal(t, 100, config.ConsulServiceCheckTimeout())
}

func TestConfigOverride(t *testing.T) {
	clientConfig := config.NewConfig()
	clientConfig.NodeID = "example-service"
	clientConfig.NodeIP = "example-ip"
	clientConfig.AppName = "example-service"
	assert.Equal(t, "example-service", config.NodeID())
	assert.Equal(t, "example-ip", config.NodeIP())
	assert.Equal(t, "example-service", config.AppName())
}



clients/consul/consul_client_test.go
package consul_test

import (
	"testing"
	"github.com/usrlocalts/consul_client/config"
	"github.com/usrlocalts/consul_client/clients/consul"
	"github.com/usrlocalts/consul_client/testutil"
	"github.com/stretchr/testify/assert"
	"reflect"
)

func TestNewConsulClient(t *testing.T) {
	clientConfig := config.NewConfig()
	consulClient, err := consul.NewConsulClient(clientConfig)
	assert.NoError(t, err)
	assert.Equal(t, reflect.TypeOf(consulClient).String(), "*consul.client")
}

func TestConsulClientWithWrongProtocolShouldReturnError(t *testing.T) {
	clientConfig := config.NewConfig()
	clientConfig.ConsulAddress = "blah://blah.com:8600"
	_, err := consul.NewConsulClient(clientConfig)
	assert.Error(t, err)
}

func TestConsulClientShouldRegisterToConsul(t *testing.T) {
	clientConfig := config.NewConfig()
	clientConfig.NodeID = "example-service"
	clientConfig.NodeIP = "example-ip"
	clientConfig.AppName = "example-service"

	lis := testutil.StartTestConsulServer()
	address := lis.Addr()
	server := testutil.MockConsulResponse("example-service", []string{"grpc"}, address)
	clientConfig.ConsulAddress = server.Listener.Addr().String()
	defer server.Close()

	consulClient, err := consul.NewConsulClient(clientConfig)
	assert.NoError(t, err)
	err = consulClient.RegisterToConsul()
	assert.NoError(t, err)
}

func TestConsulClientShouldNotRegisterToConsulWithNoServerRunningError(t *testing.T) {
	clientConfig := config.NewConfig()
	clientConfig.NodeID = "example-service"
	clientConfig.NodeIP = "example-ip"
	clientConfig.AppName = "example-service"

	consulClient, err := consul.NewConsulClient(clientConfig)
	assert.NoError(t, err)
	err = consulClient.RegisterToConsul()
	assert.Error(t, err)
}

func TestConsulClientShouldDeRegisterFromConsul(t *testing.T) {
	clientConfig := config.NewConfig()
	clientConfig.NodeID = "example-service"
	clientConfig.NodeIP = "example-ip"
	clientConfig.AppName = "example-service"

	lis := testutil.StartTestConsulServer()
	address := lis.Addr()
	server := testutil.MockConsulResponse("example-service", []string{"grpc"}, address)
	clientConfig.ConsulAddress = server.Listener.Addr().String()
	defer server.Close()

	consulClient, err := consul.NewConsulClient(clientConfig)
	assert.NoError(t, err)
	err = consulClient.DeRegisterFromConsul("12345")
	assert.NoError(t, err)
}

func TestConsulClientShouldNotDeRegisterFromConsulWithNoServerRunningError(t *testing.T) {
	clientConfig := config.NewConfig()
	clientConfig.NodeID = "example-service"
	clientConfig.NodeIP = "example-ip"
	clientConfig.AppName = "example-service"

	consulClient, err := consul.NewConsulClient(clientConfig)
	assert.NoError(t, err)
	err = consulClient.DeRegisterFromConsul("12345")
	assert.Error(t, err)
}

func TestDiscoverMultipleServicesByNameAndTag(t *testing.T) {
	mockConsulResponse := "[{\"Address\": \"127.0.0.1\", \"ServiceName\": \"example-service\", \"ServiceTags\": [], \"ServiceAddress\": \"localhost\", \"ServicePort\": 5000 }, {\"Address\": \"127.0.0.1\", \"ServiceName\": \"example-service\", \"ServiceTags\": [], \"ServiceAddress\": \"localhost\", \"ServicePort\": 5001 }]"
	server := testutil.SetupMockConsulResponse(mockConsulResponse)
	defer server.Close()

	clientConfig := config.NewConfig()
	clientConfig.NodeID = "example-service"
	clientConfig.NodeIP = "example-ip"
	clientConfig.AppName = "example-service"

	clientConfig.ConsulAddress = server.Listener.Addr().String()
	defer server.Close()

	consulClient, err := consul.NewConsulClient(clientConfig)

	exampleServices, _ := consulClient.Discover("example-service", []string{"grpc"})

	assert.NoError(t, err)
	assert.Equal(t, 2, len(exampleServices))
	assert.Equal(t, "127.0.0.1:5000", exampleServices[0])
	assert.Equal(t, "127.0.0.1:5001", exampleServices[1])
}

func TestDiscoverServiceIfAddressNotFound(t *testing.T) {
	clientConfig := config.NewConfig()
	clientConfig.NodeID = "example-service"
	clientConfig.NodeIP = "example-ip"
	clientConfig.AppName = "example-service"

	consulClient, err := consul.NewConsulClient(clientConfig)
	assert.NoError(t, err)

	exampleServices, _ := consulClient.Discover("example-service", []string{"grpc"})
	assert.Equal(t, 0, len(exampleServices))
}


clients/consul/discovery/service_discovery_test.go
package discovery_test

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/usrlocalts/consul_client/clients/consul/discovery"
	"github.com/usrlocalts/consul_client/testutil"
)

func TestDiscoverMultipleServicesByNameAndTag(t *testing.T) {
	mockConsulResponse := "[{\"Address\": \"127.0.0.1\", \"ServiceName\": \"example-service\", \"ServiceTags\": [], \"ServiceAddress\": \"localhost\", \"ServicePort\": 5000 }, {\"Address\": \"127.0.0.1\", \"ServiceName\": \"example-service\", \"ServiceTags\": [], \"ServiceAddress\": \"localhost\", \"ServicePort\": 5001 }]"
	server := testutil.SetupMockConsulResponse(mockConsulResponse)
	defer server.Close()
	sdd := discovery.New(server.Listener.Addr().String())

	exampleServices, err := sdd.Discover("example-service", []string{"grpc"})
	assert.NoError(t, err)
	assert.Equal(t, 2, len(exampleServices))
	assert.Equal(t, "127.0.0.1:5000", exampleServices[0])
	assert.Equal(t, "127.0.0.1:5001", exampleServices[1])
}

func TestDiscoverServiceIfAddressNotFound(t *testing.T) {
	mockConsulResponse := "[]"
	server := testutil.SetupMockConsulResponse(mockConsulResponse)
	defer server.Close()
	sdd := discovery.New(server.Listener.Addr().String())

	exampleServices, _ := sdd.Discover("example-service", []string{"grpc"})
	assert.Equal(t, 0, len(exampleServices))
}


testutil/mockconsul.go
package testutil

import (
	"fmt"
	"io"
	"net"
	"net/http"
	"net/http/httptest"
	"strings"
)

type MockConsul struct {
	response string
}

func (self *MockConsul) ResetHandlerResponse(response string) {
	self.response = response
}

func (self *MockConsul) GetMockServer() (*httptest.Server) {
	handler := func(w http.ResponseWriter, r *http.Request) {
		if self.response == "" {
			w.WriteHeader(http.StatusNotFound)
		} else {
			io.WriteString(w, self.response)
		}
	}

	server := httptest.NewServer(http.HandlerFunc(handler))

	return server
}

func NewMockConsul(response string) *MockConsul {
	return &MockConsul{response: response}
}

func SetupMockConsulResponse(response string) *httptest.Server {
	return NewMockConsul(response).GetMockServer()
}

func MockConsulResponse(name string, tags []string, addresses ...net.Addr) *httptest.Server {
	var response []string
	for _, address := range addresses {
		addressData := strings.Split(address.String(), ":")
		tagData := strings.Join(tags, "\",\"")
		consulData := fmt.Sprintf("{\"Address\": \"%s\", \"ServiceName\": \"%s\", \"ServiceTags\": [\"%s\"], \"ServiceAddress\": \"%s\", \"ServicePort\": %s }", addressData[0], name, tagData, addressData[0], addressData[1])
		response = append(response, consulData)
	}
	return SetupMockConsulResponse(fmt.Sprintf("[%s]", strings.Join(response, ",")))
}


clients/consul/consul_client.go
package consul

import (
	consul "github.com/hashicorp/consul/api"
	"github.com/usrlocalts/consul_client/config"
	"github.com/usrlocalts/consul_client/clients/consul/discovery"
)

type Client interface {
	RegisterToConsul() error
	DeRegisterFromConsul(string) error
	Discover(name string, tags []string) ([]string, error)
}

type client struct {
	consul           *consul.Client
	consulConf       *config.Config
	serviceDiscovery discovery.ServiceDiscovery
}

func NewConsulClient(consulConf *config.Config) (Client, error) {
	config := consul.DefaultConfig()
	config.Address = consulConf.ConsulAddress
	c, err := consul.NewClient(config)
	if err != nil {
		return nil, err
	}
	return &client{consul: c, consulConf: consulConf, serviceDiscovery: discovery.New(consulConf.ConsulAddress)}, nil
}

func (c *client) RegisterToConsul() error {
	reg := &consul.AgentServiceRegistration{
		ID:      c.consulConf.NodeID,
		Name:    c.consulConf.AppName,
		Port:    c.consulConf.Port,
		Address: c.consulConf.NodeIP,
		Tags:    c.consulConf.ConsulTags,
		Check: &consul.AgentServiceCheck{
			Interval: c.consulConf.ConsulCheckInterval,
			HTTP:     c.consulConf.ConsulCheckURL,
		},
	}

	return c.consul.Agent().ServiceRegister(reg)
}

func (c *client) DeRegisterFromConsul(id string) error {
	return c.consul.Agent().ServiceDeregister(id)
}

func (c *client) Discover(name string, tags []string) ([]string, error){
	return c.serviceDiscovery.Discover(name, tags)
}

clients/consul/discovery/service_discovery.go
package discovery

import (
	"fmt"
	"strings"

	"github.com/hashicorp/consul/api"
)

type ServiceDiscovery interface {
	Discover(name string, tags []string) ([]string, error)
}

type serviceDiscovery struct {
	consulAgentAddress string
}

func New(consulAgentAddress string) ServiceDiscovery {
	return &serviceDiscovery{consulAgentAddress: consulAgentAddress}
}

func (self serviceDiscovery) Discover(name string, tags []string) ([]string, error) {

	consulConfig := api.DefaultConfig()
	consulConfig.Address = self.consulAgentAddress
	client, err := api.NewClient(consulConfig)
	if err != nil {
		return []string{}, err
	}
	catalog := client.Catalog()
	discoveredServices, _, err := catalog.Service(name, strings.Join(tags[:], ","), &api.QueryOptions{})
	if err != nil {
		return []string{}, err
	}
	urls := make([]string, len(discoveredServices))
	for i, service := range discoveredServices {
		urls[i] = fmt.Sprintf("%s:%d", service.Address, service.ServicePort)
	}
	return urls, nil
}


config/config.go
package config

type Config struct {
	Port                      int
	ConsulAddress             string
	NodeID                    string
	NodeIP                    string
	AppName                   string
	ConsulCheckURL            string
	ConsulCheckInterval       string
	ConsulTags                []string
	ConsulServiceCheckTimeout int
}

var appConfig *Config

func load(config *Config) {
	appConfig = config
}

func Port() int {
	return appConfig.Port
}

func ConsulAddress() string {
	return appConfig.ConsulAddress
}

func NodeID() string {
	return appConfig.NodeID
}

func NodeIP() string {
	return appConfig.NodeIP
}

func AppName() string {
	return appConfig.AppName
}

func ConsulTags() []string {
	return appConfig.ConsulTags
}

func ConsulCheckURL() string {
	return appConfig.ConsulCheckURL
}

func ConsulCheckInterval() string {
	return appConfig.ConsulCheckInterval
}

func ConsulServiceCheckTimeout() int {
	return appConfig.ConsulServiceCheckTimeout
}

func NewConfig() *Config {
	newConfig := &Config{
		Port:                      3000,
		ConsulAddress:             "localhost:8500",
		NodeIP:                    "localhost",
		ConsulCheckURL:            "http://localhost:3000/",
		ConsulCheckInterval:       "3s",
		ConsulTags:                []string{"grpc"},
		ConsulServiceCheckTimeout: 100,
	}
	load(newConfig)
	return newConfig
}
